### PR TITLE
Fixes #514

### LINKED ISSUE #514: Fix email validation in StringExtensions
`isEmail` in [`StringExtensions.swift`](https://github.com/SwifterSwift/SwifterSwift/blob/3cb82030a13a40a751b60fc5c8802e31c998baea/Sources/Extensions/SwiftStdlib/StringExtensions.swift#L163) is not checking for spaces inside the email address

Example:
```swift
let email = "john dow@ap ple.com"
print(email.isEmail) // prints true even the email is not valid!
```

### PR DESCRIPTION
**Pull request by omaralbeik**
Used [RFC 5322](http://emailregex.com/) in `isEmail`, an email address regex that 99.99% works.

## Checklist
- [x] I checked the [**Contributing Guidelines**](https://github.com/SwifterSwift/SwifterSwift/blob/master/CONTRIBUTING.md) before creating this request.
- [x] New extensions are written in Swift 4.
- [x] New extensions support iOS 8.0+ / tvOS 9.0+ / macOS 10.10+ / watchOS 2.0+.
- [x] I have added tests for new extensions, and they passed.
- [x] All extensions have a **clear** comments explaining their functionality, all parameters and return type in English.
- [x] All extensions are declared as **public**.
- [x] I have added a [changelog](https://github.com/SwifterSwift/SwifterSwift/blob/master/CHANGELOG_GUIDELINES.md) entry describing my changes.

### CODE REVIEW

File: Tests/SwiftStdlibTests/StringExtensionsTests.swift [hunk 1]
```diff
@@ -77,7 +77,12 @@ final class StringExtensionsTests: XCTestCase {
     func testIsEmail() {
         XCTAssert("john@appleseed.com".isEmail)
         XCTAssertFalse("john@appleseedcom".isEmail)
-        XCTAssertFalse("johnappleseed.com".isEmail)
+		XCTAssertFalse("johnappleseed.com".isEmail)
```
Comment (guykogus) [inline Tests/SwiftStdlibTests/StringExtensionsTests.swift]: You're indenting with tabs here, while it looks like the rest uses spaces. Should we have a defined standard?

File: Sources/Extensions/SwiftStdlib/StringExtensions.swift [hunk 2]
```diff
@@ -161,9 +161,10 @@ public extension String {
 	///		"john@doe.com".isEmail -> true
 	///
 	public var isEmail: Bool {
-		// http://stackoverflow.com/questions/25471114/how-to-validate-an-e-mail-address-in-swift
-		return range(of: "[A-Z0-9a-z._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}",
-					 options: .regularExpression, range: nil, locale: nil) != nil
+		guard self == trimmingCharacters(in: .whitespacesAndNewlines) else { return false }
```
Comment (guykogus) [inline Sources/Extensions/SwiftStdlib/StringExtensions.swift]: Instead of doing this you should just add `^` and `$` to the beginning and end of the regex, respectively. Doing that to the original version of this function would have solved #514 to begin with.

File: Sources/Extensions/SwiftStdlib/StringExtensions.swift [hunk 3]
```diff
@@ -161,9 +161,10 @@ public extension String {
...
+		// http://emailregex.com/
+		let regex = "(?:[\\p{L}0-9!#$%\\&'*+/=?\\^_`{|}~-]+(?:\\.[\\p{L}0-9!#$%\\&'*+/=?\\^_`{|}~-]+)*|\"(?:[\\x01-\\x08\\x0b\\x0c\\x0e-\\x1f\\x21\\x23-\\x5b\\x5d-\\x7f]|\\\\[\\x01-\\x09\\x0b\\x0c\\x0e-\\x7f])*\")@(?:(?:[\\p{L}0-9](?:[a-z0-9-]*[\\p{L}0-9])?\\.)+[\\p{L}0-9](?:[\\p{L}0-9-]*[\\p{L}0-9])?|\\[(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?|[\\p{L}0-9-]*[\\p{L}0-9]:(?:[\\x01-\\x08\\x0b\\x0c\\x0e-\\x1f\\x21-\\x5a\\x53-\\x7f]|\\\\[\\x01-\\x09\\x0b\\x0c\\x0e-\\x7f])+)\\])"
```
Comment (guykogus) [inline Sources/Extensions/SwiftStdlib/StringExtensions.swift]: I've read many posts about people saying that email validation is not something that should be done by regex. If you really want to ensure an email is valid you need to send it an email and see that you get a success response from the mail server.
Having such an extensive regex for emails is probably useful but it can potentially be misleading. People might look at this and think that it covers all scenarios and that there will never be false positives or negatives.
My suggestion, therefore, is that we add a note to the comment of the property to make it clear that this should only be used as a _suggestion_ as to whether or not the string is an email.
Personally I'd rather this property be named something like `looksLikeAnEmail` but I think that that would be pretty useless 😝

File: Tests/SwiftStdlibTests/StringExtensionsTests.swift (hunk 1 above)
Comment (omaralbeik) [reply Tests/SwiftStdlibTests/StringExtensionsTests.swift]: 💯true, will update this

File: Sources/Extensions/SwiftStdlib/StringExtensions.swift (hunk 2 above)
Comment (omaralbeik) [reply Sources/Extensions/SwiftStdlib/StringExtensions.swift]: Adding `^` and `$` solves the problem, however the old regex fails in examples like
`email.@.email.com` 🤔

File: Tests/SwiftStdlibTests/StringExtensionsTests.swift [hunk 4]
```diff
@@ -77,7 +77,12 @@ final class StringExtensionsTests: XCTestCase {
     func testIsEmail() {
```
Comment (LucianoPAlmeida) [inline Tests/SwiftStdlibTests/StringExtensionsTests.swift]: @omaralbeik There are some valid/invalid email test cases [here](https://blogs.msdn.microsoft.com/testing123/2009/02/06/email-address-test-cases/) 
Maybe it's worth to consider adding some of them on the tests :))

File: Sources/Extensions/SwiftStdlib/StringExtensions.swift (hunk 2 above)
Comment (guykogus) [reply Sources/Extensions/SwiftStdlib/StringExtensions.swift]: Apparently that's invalid, according to the list in @LucianoPAlmeida 's comment.

File: Sources/Extensions/SwiftStdlib/StringExtensions.swift (hunk 3 above)
Comment (omaralbeik) [reply Sources/Extensions/SwiftStdlib/StringExtensions.swift]: - the purpose of this function is to serve as a local validation step for login and signup screens, to validate the address format before sending a request to server.
- I agree with you, we should add a note to inform user it is just checking the format of the string, and not checking the address itself, how about this one: **Note that this property does not perform authentication to validate the email address. It merely determines whether its format is valid for an email address.**
- About naming, I found [Micosoft](https://docs.microsoft.com/en-us/dotnet/standard/base-types/how-to-verify-that-strings-are-in-valid-email-format) is using the name `IsValidEmail`, maybe we can do so as well 😅

**Comment by omaralbeik** [comment]
cc @SD10 ☝️

File: Sources/Extensions/SwiftStdlib/StringExtensions.swift (hunk 2 above)
Comment (LucianoPAlmeida) [reply Sources/Extensions/SwiftStdlib/StringExtensions.swift]: @omaralbeik @guykogus 
Looks like that list is based [RFC 2822](https://tools.ietf.org/html/rfc2822) that is obsolete by [RFC 5322](https://tools.ietf.org/html/rfc5322#section-3.4.1), but anyways I believe this is an invalid email because  it seems like [RFC 5322](https://tools.ietf.org/html/rfc5322) explicitly only allows dots inside the local part of the mail address, not at the beginning and end.

File: Sources/Extensions/SwiftStdlib/StringExtensions.swift (hunk 3 above)
Comment (guykogus) [reply Sources/Extensions/SwiftStdlib/StringExtensions.swift]: - Agreed, I've used this kind of validation before for past projects.
- The note wouldn't be necessarily correct. Even in your new test for `email.@.email.com` you get a false positive, because according to @LucianoPAlmeida 's link of invalid email formats the trailing `.` before the `@` isn't acceptable. That was the point I was trying to make before, which I read in a number of articles, where it's seemingly impossible to have a regex that will tell you for sure if an email address is valid or invalid. So what we're doing here is trying to _verify_, rather than _validate_, the email address. So maybe the note should be something like: **Note that this property does not validate the email address against an email server. It merely attempts to determine whether its format is suitable for an email address.**
- Let's keep the name just to avoid a breaking change 😬

Comment (LucianoPAlmeida) [reply Sources/Extensions/SwiftStdlib/StringExtensions.swift]: I agree too :))
I believe rename is important if it can improve expressivity and explain better the funcionality. Also we don't need to remove `isEmail`, we can just deprecate it and remove in a future version. And add the note on documentation as well.  I like names like you guys suggested `IsValidEmail` or maybe `matchesEmailPattern`. What do you think?

Comment (SD10) [reply Sources/Extensions/SwiftStdlib/StringExtensions.swift]: The purpose of this property is to determine if some text is an email, locally, before passing it on to the server for further validation. I'm fine with the name `isEmail` or `isValidEmail`. It's an extension on `String` so I believe it's self explanatory that a Swift primitive would not be performing validation server side

Comment (omaralbeik) [reply Sources/Extensions/SwiftStdlib/StringExtensions.swift]: I see, then I'm deprecating `isEmail` and renaming it to `isValidEmail` to have better consistency with other properties like `isValidURL`

@guykogus I also updated the note 👍

File: Tests/SwiftStdlibTests/StringExtensionsTests.swift (hunk 4 above)
Comment (omaralbeik) [reply Tests/SwiftStdlibTests/StringExtensionsTests.swift]: Great, I added tests from links as well 💯

File: Tests/SwiftStdlibTests/StringExtensionsTests.swift [hunk 5]
```diff
@@ -74,10 +74,37 @@ final class StringExtensionsTests: XCTestCase {
         XCTAssertFalse("abc".isAlphaNumeric)
     }
 
-    func testIsEmail() {
-        XCTAssert("john@appleseed.com".isEmail)
-        XCTAssertFalse("john@appleseedcom".isEmail)
-        XCTAssertFalse("johnappleseed.com".isEmail)
+    func testisValidEmail() {
+        // https://blogs.msdn.microsoft.com/testing123/2009/02/06/email-address-test-cases/
```
Comment (LucianoPAlmeida) [inline Tests/SwiftStdlibTests/StringExtensionsTests.swift]: @omaralbeik Maybe we should add the untrimmed email like "     email@mail.com  " .isValidEmail test case as well. What you think?

File: Sources/Extensions/SwiftStdlib/Deprecated/SwiftStdlibDeprecated.swift [hunk 6]
```diff
@@ -39,6 +39,20 @@ public extension Bool {
...
 extension String {
 
+	#if canImport(Foundation)
+	/// SwifterSwift: Check if string is valid email format.
+	/// **Note that this property does not validate the email address against an email server. It merely attempts to determine whether its format is suitable for an email address.**.
+	///
+	///		"john@doe.com".isEmail -> true
+	///
+	@available(*, deprecated: 4.5.0, message: "Use isValidEmail instead", renamed: "isValidEmail")
+	public var isEmail: Bool {
```
Comment (guykogus) [inline Sources/Extensions/SwiftStdlib/Deprecated/SwiftStdlibDeprecated.swift]: Why did you move this to a different extension?

File: Sources/Extensions/SwiftStdlib/StringExtensions.swift [hunk 7]
```diff
@@ -157,13 +157,14 @@ public extension String {
 
 	#if canImport(Foundation)
 	/// SwifterSwift: Check if string is valid email format.
+	/// **Note that this property does not validate the email address against an email server. It merely attempts to determine whether its format is suitable for an email address.**.
```
Comment (guykogus) [inline Sources/Extensions/SwiftStdlib/StringExtensions.swift]: You can use the `Note` annotation instead. I.e.
```
///
/// - Note:
///  Note that this property does not validate the email address against an email server. It merely attempts to determine whether its format is suitable for an email address.
///
```
That way it'll appear in bold in the callout.

File: Sources/Extensions/SwiftStdlib/Deprecated/SwiftStdlibDeprecated.swift (hunk 6 above)
Comment (SD10) [reply Sources/Extensions/SwiftStdlib/Deprecated/SwiftStdlibDeprecated.swift]: I'm guessing to deprecate the old API

Comment (LucianoPAlmeida) [reply Sources/Extensions/SwiftStdlib/Deprecated/SwiftStdlibDeprecated.swift]: @guykogus We move the deprecate extensions to those files to exclude them for coverage reports :))

Comment (SD10) [reply Sources/Extensions/SwiftStdlib/Deprecated/SwiftStdlibDeprecated.swift]: Yeah it is an API breaking change that's not backwards compatible. We're not following semver properly here

Comment (guykogus) [reply Sources/Extensions/SwiftStdlib/Deprecated/SwiftStdlibDeprecated.swift]: I really need to read more carefully the name of the file! 😅

Comment (omaralbeik) [reply Sources/Extensions/SwiftStdlib/Deprecated/SwiftStdlibDeprecated.swift]: btw, I think since we used `renamed`, Xcode will suggest replacing the deprecated function with one click 😃
